### PR TITLE
WIP: Nix 3 compatibility

### PR DESCRIPTION
**Pull request by l5r**
### Description

This changes `nix-env` commands to the corresponding `nix profile` commands. It is not compatible with stable nix. How do we handle this?

Also, I should probably squash these commits and write an in-depth commit message.

See also:   #1561

### Checklist

- [ ] Change is backwards compatible.

- [ ] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by theduke** [comment]
Drive by question: where can I get some information on Nix 3?

**Comment by berbiche** [comment]
> Drive by question: where can I get some information on Nix 3?

"Nix 3.0" is officially Nix 2.4. The commit bumping it from 2.4 to 3.0 was reverted.

See https://github.com/NixOS/nix/commit/ac0e24f21b162f99c8e9437c710797820345a7a2

File: modules/home-environment.nix [hunk 1]
```diff
@@ -487,13 +487,13 @@ in
       if config.submoduleSupport.externalPackageInstall
       then
         ''
-          if nix-env -q | grep '^home-manager-path$'; then
-            $DRY_RUN_CMD nix-env -e home-manager-path
+          if nix profile info | grep '^home-manager-path$'; then
+            $DRY_RUN_CMD nix profile remove home-manager-path
           fi
```
Comment (berbiche) [inline modules/home-environment.nix:492]: Would using an `elif` branch with `nix-env -q` here work?

**Comment by mjlbach** [comment]
Have you tested this, because I'm pretty sure you can't call `nix profile install` on a bare store path:

```
❯ nix profile install /nix/store/ylbmsxkayniklphvm7ba463ynh9r1589-home-manager-path
error: --- UnimplementedError --------------------------------------------------------------------------------------------------------------------------------------------------- nix
'nix profile install' does not support argument '/nix/store/ylbmsxkayniklphvm7ba463ynh9r1589-home-manager-path'
```

**Comment by tomberek** [comment]
In case this helps anyone. This sets up two profiles, one for `nix-env`/`home-manager` and another for `nix profile`. A bash function detects this situation and injects a `--profile` parameter. This is a hack to allow experimenting with `nix profile` whilst continuing to use `home-manager`.

In home-manager:
```
programs.bash.initExtra = ''
      function nix(){                                                           
        local subcommand                                                        
        if (( "$#" == 0 )); then command nix; return; fi                        
        subcommand=$1; shift                                                    
        case $subcommand in                                                     
          profile)  command nix "$subcommand" "$@" --profile $HOME/.nix-profile-new ;;
          *)        command nix "$subcommand" "$@" ;;                           
        esac                                                                    
      }                                                                         
'';
```

In NixOS configuration:
```
  environment.profiles = [ "$HOME/.nix-profile-new" ];
```

**Pull request closed without merging**